File: .agents/skills/rain-issue-consumer-card/scripts/decrypt-card-secret.ts
/**
 * Decrypt a Rain encrypted card secret (PAN or CVC).
 *
 * CORRECTED AES-128-GCM. Rain's published Node snippet skips setAuthTag()/final() and
 * decrypts over the auth tag, leaving 16 trailing garbage bytes. This version:
 *   1. splits data into ciphertext (all but last 16 bytes) + authTag (last 16 bytes),
 *   2. calls setAuthTag(authTag),
 *   3. decrypts ONLY the ciphertext and calls final() (which verifies the tag).
 *
 * Inputs come from `GET /issuing/cards/{cardId}/secrets` (or the scoped-card response):
 *   { encryptedPan: { iv, data }, encryptedCvc: { iv, data } }
 * `secretKey` is the 32-char hex string returned by generate-session-id.
 *
 * Two implementations:
 *   - decryptSecret(data, iv, secretKey)          — Node (crypto module)
 *   - decryptSecretWebCrypto(data, iv, secretKey) — browser / Worker (Web Crypto API)
 *
 * NOTE arg order: (data, iv, secretKey) — `data` first, then `iv`.
 * SECURITY: never log the return value. Hand it to your secure surface and discard.
 */

import crypto from "node:crypto";

const HEX = /^[0-9A-Fa-f]+$/;

/** Node — corrected AES-128-GCM decrypt. */
export function decryptSecret(
  base64Data: string,
  base64Iv: string,
  secretKeyHex: string,
): string {
  if (!base64Data) throw new Error("base64Data is required");
  if (!base64Iv) throw new Error("base64Iv is required");
  if (!secretKeyHex || !HEX.test(secretKeyHex)) {
    throw new Error("secretKey must be a hex string");
  }

  const key = Buffer.from(secretKeyHex, "hex"); // 16 bytes -> AES-128
  const iv = Buffer.from(base64Iv, "base64");
  const buf = Buffer.from(base64Data, "base64");

  // Wire format is `ciphertext || 16-byte GCM tag`.
  const tag = buf.subarray(buf.length - 16);
  const ciphertext = buf.subarray(0, buf.length - 16);

  const decipher = crypto.createDecipheriv("aes-128-gcm", key, iv);
  decipher.setAuthTag(tag); // <-- missing in Rain's official snippet
  // final() verifies the tag and throws on mismatch; do NOT skip it.
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf-8");
}

/** Browser / Worker — Web Crypto handles the trailing tag automatically. */
export async function decryptSecretWebCrypto(
  base64Data: string,
  base64Iv: string,
  secretKeyHex: string,
): Promise<string> {
  if (!base64Data) throw new Error("base64Data is required");
  if (!base64Iv) throw new Error("base64Iv is required");
  if (!secretKeyHex || !HEX.test(secretKeyHex)) {
    throw new Error("secretKey must be a hex string");
  }

  const data = Uint8Array.from(atob(base64Data), (c) => c.charCodeAt(0));
  const iv = Uint8Array.from(atob(base64Iv), (c) => c.charCodeAt(0));
  const keyBytes = Uint8Array.from(
    secretKeyHex.match(/.{1,2}/g)!.map((b) => parseInt(b, 16)),
  );

  const key = await crypto.subtle.importKey(
    "raw",
    keyBytes,
    { name: "AES-GCM" },
    false,
    ["decrypt"],
  );

  // subtle.decrypt treats the last 16 bytes of `data` as the auth tag and verifies it.
  const plain = await crypto.subtle.decrypt({ name: "AES-GCM", iv }, key, data);
  return new TextDecoder().decode(plain);
}

// Self-test: round-trip a known plaintext (no live API needed).
//   npx tsx decrypt-card-secret.ts --selftest
if (typeof require !== "undefined" && require.main === module && process.argv.includes("--selftest")) {
  const secretKey = "00112233445566778899aabbccddeeff";
  const key = Buffer.from(secretKey, "hex");
  const iv = crypto.randomBytes(12);
  const known = "4111111111111111";
  const cipher = crypto.createCipheriv("aes-128-gcm", key, iv);
  const ct = Buffer.concat([cipher.update(known, "utf-8"), cipher.final()]);
  const data = Buffer.concat([ct, cipher.getAuthTag()]).toString("base64");
  const out = decryptSecret(data, iv.toString("base64"), secretKey);
  if (out !== known || out.length !== known.length) {
    throw new Error(`round-trip failed: got "${out}" (len ${out.length})`);
  }
  console.log("self-test OK: round-trip exact, no trailing bytes");
}

File: .agents/skills/rain-issue-consumer-card/scripts/decrypt-card-secret.py
"""
Decrypt a Rain encrypted card secret (PAN or CVC).

CORRECTED AES-128-GCM. The wire format is `ciphertext || 16-byte GCM tag`. This impl
splits off the last 16 bytes as the tag, decrypts only the ciphertext, and verifies the
tag (raising InvalidTag on mismatch).

Inputs come from `GET /issuing/cards/{cardId}/secrets` (or the scoped-card response):
  { "encryptedPan": {"iv","data"}, "encryptedCvc": {"iv","data"} }
`secret_key` is the 32-char hex string returned by generate_session_id.

NOTE arg order: (base64_data, base64_iv, secret_key) — data first, then iv.
SECURITY: never log the return value.

Requires: pip install cryptography
Self-test:  python decrypt_card_secret.py --selftest
"""

import base64
import binascii
import os
import sys

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def decrypt_secret(base64_data: str, base64_iv: str, secret_key_hex: str) -> str:
    if not base64_data:
        raise ValueError("base64_data is required")
    if not base64_iv:
        raise ValueError("base64_iv is required")
    if not secret_key_hex or not all(c in "0123456789abcdefABCDEF" for c in secret_key_hex):
        raise ValueError("secret_key must be a hex string")

    key = binascii.unhexlify(secret_key_hex)  # 16 bytes -> AES-128
    iv = base64.b64decode(base64_iv)
    buf = base64.b64decode(base64_data)        # ciphertext || 16-byte tag

    # cryptography's AESGCM.decrypt expects (ciphertext_with_tag_appended); it slices the
    # trailing 16-byte tag and verifies it automatically — the correct GCM behavior.
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, buf, None)
    return plaintext.decode("utf-8")


if __name__ == "__main__":
    if "--selftest" in sys.argv:
        secret_key = "00112233445566778899aabbccddeeff"
        key = binascii.unhexlify(secret_key)
        iv = os.urandom(12)
        known = "4111111111111111"
        ct_with_tag = AESGCM(key).encrypt(iv, known.encode(), None)  # appends the tag
        data = base64.b64encode(ct_with_tag).decode()
        out = decrypt_secret(data, base64.b64encode(iv).decode(), secret_key)
        assert out == known and len(out) == len(known), f'round-trip failed: {out!r}'
        print("self-test OK: round-trip exact, no trailing bytes")
